### PR TITLE
Column takeset

### PR DESCRIPTION
**Pull request by jeromekelleher**
Provide x_table_takeset_columns

Part of #2016, #1977, #537, #538

This is my initial outline for discussion, and completely untested. What do you think @benjeffery?

### CODE REVIEW

**Review by benjeffery — COMMENTED** [review]
Looks good, shame we have to duplicate some logic but a separate function is the way.

File: c/tskit/tables.c [hunk 1]
```diff
@@ -1558,6 +1678,25 @@ tsk_individual_table_load(tsk_individual_table_t *self, kastore_t *store)
  * node table
  *************************/
 
+static void
+tsk_node_table_free_columns(tsk_node_table_t *self)
+{
+    tsk_safe_free(self->flags);
```
Comment (jeromekelleher) [inline c/tskit/tables.c:1684]: oops, need to set these to NULL

**Comment by benjeffery** [comment]
I've rebased to master, trying this out with the kastore changes.

**Comment by benjeffery** [comment]
> Great - can you ping me when you'd like some input here please? (Would it be simpler to start your own PR or is it the same thing to just push to my branch? I don't mind)

My plan is to complete the proposal and tests for the IndividualTable, then get feedback. It's almost there so later today. Will push to this one if that's ok!

**Comment by benjeffery** [comment]
@jeromekelleher Had to move things about to get errors to happen before memory is changed. I think this is worth a review now.

**Comment by jeromekelleher** [comment]
LGTM! I've had a good think about how we might avoid the duplication of NULL logic between here and ``append_columns`` but I can't see how it could be done (at least not in a way that would be worse than the code duplication). Press on with this pattern would be my vote.

**Comment by benjeffery** [comment]
Great, I have commenced the great pasting ritual.

**Comment by benjeffery** [comment]
@jeromekelleher Think this is ready to review now!

**Comment by benjeffery** [comment]
Just realised we also need takeset indexes too!

File: c/tskit/tables.c (hunk 1 above)
Comment (benjeffery) [reply c/tskit/tables.c:1689]: Not sure I'm clear on why?

**Comment by benjeffery** [comment]
@jeromekelleher Ok! I think we are good here, I've used these methods over at #2143 successfully.

**Comment by benjeffery** [comment]
Gah, missing coverage on all the ragged-column-contents error branches. Will add that in.

**Comment by benjeffery** [comment]
Ok, tests added. I think all the remaining uncovered branches are only triggerable by malloc errors.

File: c/tskit/tables.c [hunk 2]
```diff
@@ -629,6 +629,81 @@ takeset_string(char **str, tsk_size_t *len, char *new_str, const tsk_size_t new_
...
+{
+    int ret = 0;
+    if ((data == NULL) != (offset == NULL)) {
+        ret = TSK_ERR_BAD_PARAM_VALUE;
+        goto out;
+    }
+    if (data != NULL) {
+        ret = check_offsets(num_rows, offset, 0, false);
+        if (ret != 0) {
+            goto out;
```
Comment (jeromekelleher) [inline c/tskit/tables.c:658]: This can be triggered by constructing bad offsets, can't it?

File: c/tskit/tables.c (hunk 1 above)
Comment (jeromekelleher) [reply c/tskit/tables.c:1689]: I'm not either! Let's forget about it.

File: c/tskit/tables.c [hunk 3]
```diff
@@ -2423,6 +2611,40 @@ tsk_edge_table_set_columns(tsk_edge_table_t *self, tsk_size_t num_rows,
...
+
+    tsk_edge_table_free_columns(self);
+    self->num_rows = num_rows;
+    self->max_rows = num_rows;
+    self->left = left;
+    self->right = right;
+    self->parent = parent;
+    self->child = child;
+
+    ret = takeset_ragged_column(num_rows, metadata, metadata_offset,
```
Comment (jeromekelleher) [inline c/tskit/tables.c:2639]: How does this interact with NO_EDGE_METADATA?

File: c/tskit/tables.c [hunk 4]
```diff
@@ -3784,6 +4061,70 @@ tsk_mutation_table_append_columns(tsk_mutation_table_t *self, tsk_size_t num_row
...
+    self->max_rows = num_rows;
+    self->site = site;
+    self->node = node;
+
+    ret = takeset_optional_id_column(num_rows, parent, &self->parent);
+    if (ret != 0) {
+        goto out;
+    }
+    if (time == NULL) {
+        /* Flags defaults to all zeros if not specified. The column is often
```
Comment (jeromekelleher) [inline c/tskit/tables.c:4098]: Comments incorrect

File: c/tskit/tables.c [hunk 5]
```diff
@@ -10074,6 +10525,14 @@ tsk_table_collection_set_metadata(
         &self->metadata, &self->metadata_length, metadata, metadata_length);
 }
 
+int
```
Comment (jeromekelleher) [inline c/tskit/tables.c:10528]: Need tests to cover this

File: c/tskit/tables.c [hunk 6]
```diff
@@ -768,13 +843,6 @@ tsk_reference_sequence_takeset_data(
     return takeset_string(&self->data, &self->data_length, data, data_length);
 }
 
-int
```
Comment (jeromekelleher) [inline c/tskit/tables.c:771]: What's the rationale for removing these - are we not using them in the ``load`` codepath and you figure it's just not worthwhile? (I don't mind, just checking)

**Review by jeromekelleher — COMMENTED** [review]
Looks great! Thanks for slogging through this @benjeffery.

I think we just need to add a few tests to cover the ``check_ragged_column`` calls on the input. It should be straightforward enough to do by setting ``x_col_offset`` = -1 (or something). That will bump up the test coverage quite a bit.

File: c/tskit/tables.c (hunk 6 above)
Comment (benjeffery) [reply c/tskit/tables.c:771]: It was to make them consistent with the other tables where schemas and other "short" strings are don't have a `takeset`. I really didn't see the point in adding a `takeset_metadata_schema` to all the tables when I can't ever imagine it being over 1MB. (Famous last words!)

File: c/tskit/tables.c (hunk 4 above)
Comment (benjeffery) [reply c/tskit/tables.c]: Fixed

File: c/tskit/tables.c (hunk 6 above)
Comment (jeromekelleher) [reply c/tskit/tables.c:771]: Sure - can always add it later.

File: c/tskit/tables.c (hunk 2 above)
Comment (benjeffery) [reply c/tskit/tables.c:658]: Tests added!

File: c/tskit/tables.c (hunk 5 above)
Comment (benjeffery) [reply c/tskit/tables.c:10531]: Done!

**Comment by jeromekelleher** [comment]
Can you ping me when this is ready for a last look please @benjeffery?

**Comment by benjeffery** [comment]
@jeromekelleher Ok! Coverage looks a lot better.

**Review by benjeffery — APPROVED** [review]

**Pull request merged**